elastic/logstash
# Field Reference: handle special characters

## Release notes

 - Fixes a long-standing issue where setting a field whose name includes characters that are reserved in the FieldReference syntax (`[` and `]`) had undefined behaviour that resulted in pipeline crashes or the silent truncation of the field name.
 - Introduces two experimental opt-in methods for referencing fields whose names include characters that are reserved in the FieldReference syntax

## What does this PR do?

This PR has three commits that tell a story:

1. Demonstrate previously-unvalidated edge-cases in which FieldReference-special characters in field names result in undesired behaviour:
   | Category | Behaviour | Example |
   | --- | ---- | ---- |
   | Valid field reference | truncate path, leaving only key | `[deeply][nested][square-brackets][field]` -> `field` |
   | Invalid field reference | rejected input, potentially crash | `log[WARN]` |
3. Fixes the Event API's `ConvertedMap` internals so that data structures with fields whose names include FieldReference-special characters (currently `[` and `]`) correctly behave as expected. This has the down-side of allowing fields on an event that _cannot_ be referenced using the FieldReference syntax.
4. Introduces two _EXPERIMENTAL_ opt-in ways of being able to reference fields with special characters in their names. 

## Why is it important/What is the impact to the user?

- Allows plugins that handle as-given data-structures (like the JSON codec) to correctly create events from the events as-given, even if those events contain FieldReference-special characters.
- Provides an experimental path forward for referencing fields whose names contain FieldReference-special characters

## Checklist




## How to test this PR locally

With the following `pipeline.conf`, we create an event from JSON data that contains a field named `[bracketed][field]`, and attempt to address that field in a variety of ways (field reference literals, sprintf):

~~~
input { generator { count => 1 codec => json message => '{"[bracketed][field]": "okay"}' } }
filter {
  if [bracketed][field] {
    mutate { add_tag => "MATCH-literal" }
  }
  if [%5Bbracketed%5D%5Bfield%5D] {
    mutate { add_tag => "MATCH-percent" }
  }
  if [&#91;bracketed&#93;&#91;field&#93;] {
    mutate { add_tag => "MATCH-ampersand" }
  }
  mutate {
    replace => {
      "sprintf" => "LITERAL:~%{[bracketed][field]}~ PERCENT:~%{%5Bbracketed%5D%5Bfield%5D}~ AMPERSAND:~%{&#91;bracketed&#93;&#91;field&#93;}~"
    }
  }
}
output { stdout { codec => rubydebug } }
~~~

Execute each of the following to observe behaviour:

~~~
bin/logstash -f pipeline.conf
~~~
~~~
bin/logstash -f pipeline.conf --field-reference-escape-style=PERCENT
~~~
~~~
bin/logstash -f pipeline.conf --field-reference-escape-style=AMPERSAND
~~~

## Related issues

Resolves: https://github.com/elastic/logstash/issues/13606
Resolves: https://github.com/elastic/logstash/issues/11608
Supersedes: https://github.com/elastic/logstash/pull/13479
